### PR TITLE
Vulkan backend

### PR DESCRIPTION
**Pull request by kekekeks**
Adds Vulkan backend. 

Note that importing DXGI handles for interop purposes isn't supported yet, only opaque vulkan handles are supported

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0039014-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

File: src/Avalonia.Vulkan/VulkanExternalObjectsFeature.cs [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    {
+        _pool.FreeFinishedCommandBuffers();
+        if (!SupportedImageHandleTypes.Contains(handle.HandleDescriptor))
+            throw new NotSupportedException();
+        
+
+        return new ImportedImage(_context, _pool, handle, properties);
+    }
+
+    [DllImport("/home/kekekeks/Projects/pinvokewtf/cmake-build-debug/libpinvokewtf.so")]
```
Comment (TETYYS) [inline src/Avalonia.Vulkan/VulkanExternalObjectsFeature.cs:122]: debug code

**Review by jmacato — APPROVED** [review]

File: src/Avalonia.Vulkan/Avalonia.Vulkan.csproj [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFrameworks>net6.0;netstandard2.0</TargetFrameworks>
```
Comment (maxkatz6) [inline src/Avalonia.Vulkan/Avalonia.Vulkan.csproj:4]: Need to update target frameworks here as we did in other projects.

File: src/Avalonia.Vulkan/IVulkanContextExternalObjectsFeature.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+using System;
+using System.Collections.Generic;
+using Avalonia.Metadata;
+using Avalonia.Platform;
+using Avalonia.Rendering.Composition;
+
+namespace Avalonia.Vulkan;
+
+[Unstable]
```
Comment (maxkatz6) [inline src/Avalonia.Vulkan/IVulkanContextExternalObjectsFeature.cs:9]: We should start marking Unstable attribute with a specific reason why. Here and in other places.

File: src/Avalonia.Vulkan/IVulkanDevice.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    public IEnumerable<string> EnabledExtensions { get; }
+}
+
+[NotClientImplementable]
+public interface IVulkanPlatformGraphicsContext : IPlatformGraphicsContext
+{
+    IVulkanDevice Device { get; }
+    IVulkanInstance Instance { get; }
+    internal VulkanInstanceApi InstanceApi { get; }
+    internal VulkanDeviceApi DeviceApi { get; }
```
Comment (maxkatz6) [inline src/Avalonia.Vulkan/IVulkanDevice.cs:34]: Does mix of internal and public interface members work on netstandard? I am surprised it's compiled.

Comment (maxkatz6) [reply src/Avalonia.Vulkan/IVulkanDevice.cs:34]: I guess it works.

File: src/Avalonia.Vulkan/Interop/VulkanCommandBufferPool.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            next.Dispose();
+        }
+    }
+
+    public void Dispose()
+    {
+        FreeUsedCommandBuffers();
+        
+        if (_handle.Handle != 0)
+            _context.DeviceApi.DestroyCommandPool(_context.DeviceHandle, _handle, IntPtr.Zero);
```
Comment (maxkatz6) [inline src/Avalonia.Vulkan/Interop/VulkanCommandBufferPool.cs:50]: I assume it's expected to skip ThrowOnError here

File: src/Avalonia.Vulkan/Interop/VulkanDebugLogger.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            VkDebugUtilsMessageSeverityFlagsEXT.VK_DEBUG_UTILS_MESSAGE_SEVERITY_ERROR_BIT_EXT => LogEventLevel.Error,
+            VkDebugUtilsMessageSeverityFlagsEXT.VK_DEBUG_UTILS_MESSAGE_SEVERITY_WARNING_BIT_EXT =>
+                LogEventLevel.Warning,
+            VkDebugUtilsMessageSeverityFlagsEXT.VK_DEBUG_UTILS_MESSAGE_SEVERITY_VERBOSE_BIT_EXT =>
+                LogEventLevel.Verbose,
+            VkDebugUtilsMessageSeverityFlagsEXT.VK_DEBUG_UTILS_MESSAGE_SEVERITY_INFO_BIT_EXT => LogEventLevel
+                .Information,
+            _ => LogEventLevel.Information
+        };
+        if (Logger.TryGet(level, "Vulkan", out var logger))
```
Comment (maxkatz6) [inline src/Avalonia.Vulkan/Interop/VulkanDebugLogger.cs:29]: Would be better to move "Vulkan" to a LogArea enum-class, so we have a self-documenting logger class.

File: src/Skia/Avalonia.Skia/Gpu/Vulkan/VulkanSkiaExternalObjectsFeature.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            };
+            using var renderTarget = new GRBackendRenderTarget(_properties.Width, _properties.Height, 1, imageInfo);
+            using var surface = SKSurface.Create(_gpu.GrContext, renderTarget,
+                _properties.TopLeftOrigin ? GRSurfaceOrigin.TopLeft : GRSurfaceOrigin.BottomLeft,
+                _properties.Format == PlatformGraphicsExternalImageFormat.R8G8B8A8UNorm
+                    ? SKColorType.Rgba8888
+                    : SKColorType.Bgra8888, SKColorSpace.CreateSrgb());
+            var image = surface.Snapshot();
+            _gpu.GrContext.Flush();
+            //surface.Canvas.Flush();
```
Comment (maxkatz6) [inline src/Skia/Avalonia.Skia/Gpu/Vulkan/VulkanSkiaExternalObjectsFeature.cs:104]: I assume it's a debug code, and surface disposal will flush canvas as well.

File: src/Windows/Avalonia.Win32/Win32GlManager.cs [hunk 8]
```diff
@@ -62,6 +64,13 @@ static class Win32GlManager
                     return wgl;
                 }
             }
+
+            if (renderingMode == Win32RenderingMode.Vulkan)
+            {
+                var vulkan = VulkanSupport.TryInitialize(AvaloniaLocator.Current.GetService<VulkanOptions>() ?? new());
+                if (vulkan != null)
+                    return vulkan;
+            }
```
Comment (maxkatz6) [inline src/Windows/Avalonia.Win32/Win32GlManager.cs:73]: So, no compositor support for now?

**Pull request merged**